### PR TITLE
[ENH, MRG] Add compute_csd function

### PR DESCRIPTION
**Pull request by alexrockhill**
Part of https://github.com/mne-tools/mne-python/pull/10920.

This uses power to compute the cross-spectral density. It would be nice if this could be done on complex numbers but that's causing issues with positive semi-definite matrices. It might be nice to just go ahead with this version and then improve it later since it matches how other functions like `csd_morlet` do CSD and then modify all the functions to be consistent in a followup PR.

### CODE REVIEW

**Comment by alexrockhill** [comment]
Failed tests are because this depends on https://github.com/mne-tools/mne-python/pull/10979

**Comment by alexrockhill** [comment]
@wmvanvliet if you have a minute to review this one too that would be really great :)

**Comment by alexrockhill** [comment]
Needs to check that it gives the same results as `csd_morlet` etc., I think there's a scaling issue.

**Comment by alexrockhill** [comment]
Okay, so this passes tests locally that calling `tfr_morlet(epochs)` and then `compute_csd(epochs_tfr)` is the same as `csd_morlet(epochs)` by changing it to match. But, a couple important questions remain:

- Does this code work for `csd_fourier` and `csd_multitaper`? I wasn't able to check because the parameterization is currently different (`tfr_multitaper` takes `n_cycles` whereas `csd_multitaper` takes an `fmin` and `fmax`), @drammock do you know if there's a backward compatible way to match the `fmin`/`fmax` parameterization with the `freqs`/`n_cycles` parameterization?
- Would this theoretically work with Hilbert transformed data? Maybe this is an empirical question to compare with the filtered raw data approach, tagging @britta-wstnr as I think she might be interested
- There are many different methods in `mne.compute_covariance`, can we allow those as well? It takes `epochs` with dimension (`n_epochs`, `n_channels`, `n_times`) so it would need to take 4D input or somehow handle frequencies. The way that I had it in the previous commits returned something sensible but didn't match `csd_morlet` even when normalized so I think there's some scaling going on that may make things different but it would be interesting to know if those give better results.

I think question 1 should be addressed before merging but questions 2 and 3 are more long-term trying to figure out how best to do this kind of thing.

**Comment by alexrockhill** [comment]
I took a bit more of a look at it appears that `tfr_mutlitaper` doesn't even accept an `output` argument to get complex data. This was easy to add (you have to average across the tapers which I'm not sure works really well for phase though) but did not give quantitatively similar results using the same frequencies (you still have the `freqs` + `n_cycles` parameterization vs the `fmin`/`fmax` parameterization) so I think it does not work. Up to people who maintain `csd` functions more whether to go ahead and merge knowing that it only works for Morlet or add a warning that it only works for `EpochsTFR` derived from `tfr_morlet` or dig into the multitaper more or what. This is a nice efficiency improvement for applying DICS to time-frequency epochs though so it would be nice to get merged at some point.

**Comment by alexrockhill** [comment]
Okay to merge? This follows the definition of CSD here

image

(From @britta-wstnr et al. 's 2022 Neuroimage paper and pointed out by her)

File: mne/time_frequency/csd.py
Comment (agramfort) [inline mne/time_frequency/csd.py:1354]: kwargs are not used in the function.

File: mne/time_frequency/csd.py
Comment (agramfort) [inline mne/time_frequency/csd.py:1379]: to avoid the for loop I would compute the full CSD matrix and use tri_indices to only store the upper part.

Comment (alexrockhill) [reply mne/time_frequency/csd.py:1379]: I tried to refactor this but it's a tricky einsum because you need to preserve the frequency dimension but average across the time dimension so you could loop over frequencies to do that but it's pretty much just as inelegant. Maybe someone else could puzzle it out but I think this is a lot harder than it looks.

**Review by agramfort — APPROVED** [review]
@drammock @larsoner @wmvanvliet ok for you?

File: mne/time_frequency/csd.py
Comment (larsoner) [reply mne/time_frequency/csd.py]: Done in 25e3c82be9

File: doc/time_frequency.rst
Comment (larsoner) [inline doc/time_frequency.rst:25]: For consistency with `csd_fourier`, `csd_multitaper`, and `csd_morlet`, maybe `csd_tfr` would be better here?

File: mne/time_frequency/csd.py
Comment (larsoner) [inline mne/time_frequency/csd.py:1393]: Does `EpochsTFR` really not have a `.info` attribute? If it has it, then projs should be in `epochs_tfr.info['projs']` already

File: mne/time_frequency/tests/test_csd.py
Comment (larsoner) [inline mne/time_frequency/tests/test_csd.py]: I didn't like this `atol` as it's a bit high. 1e-6 data will have covariance values of 1e-12 (squaring), and for CSD it might be even lower (power within a band). Locally at least it worked with no `atol` param, which should then use the defaults of `atol=0, rtol=1e-7` which seems cleaner to me. Let's see if it comes back green....

**Review by larsoner — CHANGES_REQUESTED** [review]
Just marking as changes requested until my comments above are addressed. CI failure is unrelated and should be fixed by #11034 eventually

File: mne/time_frequency/csd.py
Comment (alexrockhill) [reply mne/time_frequency/csd.py:1393]: It does have an info, I thought when I checked there wasn't a projs key in info which I thought was odd... I'll double check

Comment (larsoner) [reply mne/time_frequency/csd.py:1393]: You might need to do `info.get('projs', None)` because I think sometimes the key might just not exist. But if you compute on an instance like one from `sample` that *does* have projs and this ends up as `None` then it's a problem...

File: doc/time_frequency.rst
Comment (alexrockhill) [reply doc/time_frequency.rst:25]: I'm not too particular about it, sounds good.

File: mne/time_frequency/csd.py
Comment (alexrockhill) [reply mne/time_frequency/csd.py:1393]: That's probably what happened, it was I think made from scratch so that's probably why it didn't even have a `projs` field.

**Comment by alexrockhill** [comment]
Thanks for the review @larsoner, I added those changes. Now, https://github.com/mne-tools/mne-python/blob/main/mne/time_frequency/csd.py#L1344-L1345 is done the old none-einsum way (thanks for figuring that out by the way, I was pretty close actually but just didn't get the right combination of the first index being the only different one, I had the first and second and both the same but now that you wrote the correct one, it seems so easy) but it's done per channel so it's a bit more refactoring than should probably go in this PR.

**Comment by alexrockhill** [comment]
Looks good to go unless I missed anything

**Comment by larsoner** [comment]
Thanks @alexrockhill !

**Pull request merged**
